spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -3079,6 +3078,26 @@ public String toString() {
 					+ "\n]";
 		}
 
+		private OffsetAndMetadata createOffsetAndMetadata(long offset) {
+			final OffsetAndMetadataProvider metadataProvider = this.containerProperties.getOffsetAndMetadataProvider();
+			return metadataProvider == null
+					? new OffsetAndMetadata(offset)
+					: metadataProvider.provide(new ConsumerAwareListenerMetadata(), offset);
```

Comment:
No need to create a new one each time; since it is immutable - it can be a final field.

Also the null check can be avoided given that you already have a default implementation and you could add an `Assert.notNull` on the setter.

You could also add the provider as a final field.

Reply:
Code as been reworked as suggested.
I tried to keep the same approach than the callback at properties level. If you really prefer, I can create a default provider at property and  consider it will never be null using the assert but this may be weird to always see the default when rendering the properties...